### DISCUSSION TITLE
Sphere not staying still after fall

### DISCUSSION THREAD

**Discussion by LeXXik** [jrouwe/JoltPhysics.js#213] [Q&A] [answered]
Hmm, I am trying to investigate an issue, where a sphere spawned above ground and simply falling by gravity doesn't stay still after landing. This is similar to my previous #169, but that time it was my error, where I was accidentally applying additional forces.

This time, I simply create a sphere (r 0.10) and drop it on a box ground (he 200x1x200). No forces applied. Sometimes it stays in place after landing, but most often it rolls in some random direction.

Any ideas what may be causing this? Or is this some inherent behavior of the solver? I tried earlier Jolt versions, but that didn't seem to affect anything.

https://github.com/user-attachments/assets/c5e53b76-275e-495a-8450-e7d10c63a27b

Edit: 
~~Looks like a larger sphere radius "fixes" it. The ball stays in place after landing.~~ Nope. Still rolls.

**Comment by jrouwe** [comment, marked as answer]
I've tried to reproduce your issue in an example:

```HTML
<!DOCTYPE html>
<html lang="en">
	<head>
		<title>JoltPhysics.js demo</title>
		<meta charset="utf-8">
		<meta name="viewport" content="width=device-width, user-scalable=no, minimum-scale=1.0, maximum-scale=1.0">
		<link rel="stylesheet" type="text/css" href="style.css">
	</head>
	<body>
		<div id="container">Loading...</div>
		<div id="info">JoltPhysics.js sphere demo</div>

		<script src="js/three/three.min.js"></script>
		<script src="js/three/OrbitControls.js"></script>
		<script src="js/three/WebGL.js"></script>
		<script src="js/three/stats.min.js"></script>
		<script src="js/example.js"></script>

		<script type="module">
			// In case you haven't built the library yourself, replace URL with: https://www.unpkg.com/jolt-physics/dist/jolt-physics.wasm-compat.js
			import initJolt from './js/jolt-physics.wasm-compat.js';

			initJolt().then(function (Jolt) {
				// Initialize this example
				initExample(Jolt, null);

				let convex_radius = 0; // with this it is ok: 0.05

				// Create box
				var shape = new Jolt.BoxShape(new Jolt.Vec3(200, 1, 200), convex_radius, null);
				var creationSettings = new Jolt.BodyCreationSettings(shape, new Jolt.RVec3(0, -1, 0), Jolt.Quat.prototype.sIdentity(), Jolt.EMotionType_Static, LAYER_NON_MOVING);
				let body = bodyInterface.CreateBody(creationSettings);
				addToScene(body, 0xc7c7c7);
				
				// Create sphere
				shape = new Jolt.SphereShape(0.1);
				creationSettings = new Jolt.BodyCreationSettings(shape, new Jolt.RVec3(0, 10, 0), Jolt.Quat.prototype.sIdentity(), Jolt.EMotionType_Dynamic, LAYER_MOVING);
				creationSettings.mAllowSleeping = false; // and without this it is also ok
				body = bodyInterface.CreateBody(creationSettings);
				addToScene(body, 0x00ffff);
			});

		</script>
	</body>
</html>
```

I can reproduce the issue, but only if I set mAllowSleeping = false or when I give the box a very small convex radius.

I think the issue is that the box is very large compared to the sphere, so the collision detection algorithm has some floating point precision error which results in motion.

**Reply by LeXXik** [reply]
Hmm, right. I do have convex radius at 0.05 for the box as well. So, this takes place in actual game, where the dimensions difference is less striking:

```
ball r 0.5385
floor he 28.46 x 1 x 3.27
```

It stands still for a bit and then starts moving. This can be mitigated by allowing it to fall asleep, as you mentioned. I suppose that would be the only solution at this point?

**Reply by jrouwe** [reply]
I think you always want bodies to go to sleep. Not doing that is a major performance bottleneck.

**Reply by LeXXik** [reply]
Very true, in this particular case, the body doesn't even need to be constantly awake. I do use it (prevent sleeping) on rare occasions, though. Albeit I can't remember when I needed it last. However, only when there is only 1 or very few dynamic bodies, so the performance is not an issue. We also step physics at fixed step (1/30), so we don't hog every frame with simulations. Nowadays even phones run at 120 fps, which is too much of a detail for something crude like a ball bouncing around a room. I am getting carried away :) I will accept your answer as an answer. Thanks!

**Reply by jrouwe** [reply]
I can't recommend running the simulation at 1/30th, at that rate the artifacts really start to show. I would go for 1/60th.

**Reply by LeXXik** [reply]
Yep, we do bump to 1/60, when there are quite a few dynamic bodies interactions. Then collisions show some artifacts. But, unless artifacts show up, we keep at 1/30. Some games don't even use dynamic bodies, just a few static and kinematic bodies moving around. There is no need for 1/60 in that case.

Edit:
We target mobile phones, so every physics step is a battery drain, thus we try to keep them low.
